File: C++/91. DecodeWays.cpp
// Name: Nikita
// Github id: geekygirl8

class Solution {
public:
    int dp[105];
    
    int solve(int pos,string &s){
        if(pos==s.size()){
            return 1;
        }
        int num = s[pos]-'0';
        int &ret=dp[pos];
        if(ret!=-1){
            return ret;
        }
        ret=0;
        if(num!=0 && num<=26){
            ret=ret+solve(pos+1,s);
        }
        if(num!=0 && pos+1<s.size()){
            num = num*10 + (s[pos+1]-'0');
            if(num!=0 && num<=26){
                ret=ret+solve(pos+2,s);
            }
        }
        return ret;
    }
    
    int numDecodings(string s) {
        memset(dp,-1,sizeof dp);
        return solve(0,s);
    }
};

File: Python/Leetcode/91. DecodeWays.py
//Name: Nikita
// Github id: geekygirl8

class Solution:
    def numDecodings(self, s: str) -> int:
        # Use mapping to define the set
        mapping = set()
    
# Now we define the range and start map

        for i in range(1, 27):
            mapping.add(str(i))
        return self.break_string(s, 0, mapping, {})

    # After getting the requried set

    # Now for thhe next step we have to first seperate the data into each components so we add break_string function

    def break_string(self, s, count, mapping, mem):

    # After breaking the set (ie.,array) into its seperate indidvisuals , it should start decoding with the given grouped map into letters 

        if s == "":
            return count + 1

        ans = 0
        if s not in mem:
            for i in range(len(s)):

                if s[:i+1] in mapping:
                    ans += self.break_string(s[i+1:], count, mapping, mem)

            mem[s] = ans

        return mem[s]
